src/relay/op/tensor/reduce.cc

```diff
@@ -148,29 +158,51 @@ Array<Array<Layout>> ReduceInferCorrectLayout(const Attrs& attrs,
     tvm::Array<tvm::Integer> new_r_axes;
     std::string new_layout_string = "";
     int axis_index = 0;
-    for (auto iter_var : new_in_layouts[0]->axes) {
+    for (auto iter_var : layout->axes) {
       const auto& layout_axis = LayoutAxis::Get(iter_var);
       const std::string& layout_dim = layout_axis.name();
       if (old_r_dims.count(layout_dim)) {
         new_r_axes.push_back(tvm::Integer(axis_index));
       }
       // Collect only the primal axis.
       if (layout_axis.IsPrimal()) {
-        new_layout_string += layout_dim;
+        if (!old_r_dims.count(layout_dim)) {
+          new_layout_string += layout_dim;
+        }
         axis_index++;
       }
     }
 
     // 3) Set the new axis and layout.
-    ret = Layout(new_layout_string);
+    return std::make_tuple(params->keepdims ? stripe(layout) : Layout(new_layout_string),
+                           new_r_axes);
```

Comment:
It seems to me that `infer` could be further improved, since the desired layouts of two branches (keepdims or not) are totally different. Could we make it like the following, so that you don't need to traverse the layout twice when keepdims is true?

```c
std::string new_layout_string = "";
for (auto iter_var : layout->axes) {
  // ...
  if (layout_axis.IsPrimal()) {
    if (params->keepdims || !old_r_dims.count(layout_dim)) {
      new_layout_string += layout_dim;
    }
    axis_index++;
  }
}
return std::make_tuple(new_layout_string, new_r_axes);
```

If the above solution works, we could further consider merging `infer` and `stripe` to reduce duplicated logic:

```c
auto infer = [&](const Layout& layout, const bool keepdims) { /* ... */ };
// ...
// Origin: std::tie(inferred_out, new_r_axes) = infer(new_in_layouts[0]);
// Origin: inferred_in = stripe(new_in_layouts[0]);
std::tie(inferred_out, new_r_axes) = infer(new_in_layouts[0], params->keepdims);
std::tie(inferred_in, std::ignore) = infer(new_in_layouts[0], false);
```